### PR TITLE
Fix Issue 19754 - Re-apply #9505 and fix it, making isLvalue() logic more restrictive wrt. literals

### PR DESCRIPTION
**Pull request by kinke**
These issues have surfaced with #9505 (mainly with non-DMD backends).

### CODE REVIEW

File: src/dmd/optimize.d [hunk 1]
```diff
@@ -306,9 +306,9 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
             return expOptimize(e.e1, flags);
         }
 
-        bool binOptimize(BinExp e, int flags)
+        bool binOptimize(BinExp e, int flags, bool keepLhsLvalue = false)
```
Comment (Basile-z) [inline src/dmd/optimize.d]: No default parameters ([rule 12](https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md#dmd-best-practices))

Comment (kinke) [reply src/dmd/optimize.d]: There are 14 other call sites in this file I'd rather not touch.

**Comment by kinke** [comment]
https://github.com/dlang/dmd/blob/master/test/fail_compilation/test6883.d doesn't fail anymore.

File: src/dmd/optimize.d (hunk 1 above)
Comment (Basile-z) [reply src/dmd/optimize.d]: Let's say it's okay since it's not a new func.

**Comment by kinke** [comment]
In the 3rd commit, I tried to tentatively optimize the `IndexExp` even for `keepLvalue = true` and to 'undo' it in case it yields no lvalue (no idea if it ever would) - because that's where the compile-time errors for trivial out-of-bounds cases are generated (as part of calling `Index()` from `constfold.d`). The expected errors are back, but now twice, as each assignment is `optimize()`d twice - once as part of AssignExp-sema, and then again as ExpStatement-sema. Apparent hacks seem to be surfacing...

**Comment by kinke** [comment]
Just stumbling from one problem to the next, I'll probably give up on this. I'd suggest reverting #9505 for the time being, it has opened a can of worms and doesn't work as intended. Pinging @ibuclaw.

**Comment by kinke** [comment]
After much trial and error, there seems to be light at the end of the tunnel.

The Phobos compile errors are due to some `refRange(&["foobar"][0])...` instead of `string s = "foobar"; refRange(&s)...`. To me, it seems like a questionable shortcut to save a line; the original testcase in the associated [issue report](https://issues.dlang.org/show_bug.cgi?id=18657) uses the explicit 2nd version.

File: src/dmd/expression.d [hunk 2]
```diff
@@ -4668,12 +4668,14 @@ extern (C++) final class DotVarExp : UnaExp
 
     override bool isLvalue()
     {
-        return true;
+        return e1.op != TOK.structLiteral;
```
Comment (Basile-z) [inline src/dmd/expression.d]: The problem with structLiteral is not that wether they are Lvalue or not it's rather that when they are it's not mutable.

File: test/fail_compilation/b20011.d [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/b20011.d(23): Error: `S1(cast(ubyte)0u).member` is not an lvalue and cannot be modified
+fail_compilation/b20011.d(26): Error: `S2(null).member` is not an lvalue and cannot be modified
+fail_compilation/b20011.d(27): Error: `S2(null).member` is not an lvalue and cannot be modified
+fail_compilation/b20011.d(31): Error: `U1(cast(ubyte)0u, ).m2` is not an lvalue and cannot be modified
+fail_compilation/b20011.d(35): Error: function `b20011.main.assignableByRef(ref ubyte p)` is not callable using argument types `(ubyte)`
+fail_compilation/b20011.d(35):        cannot pass rvalue argument `S1(cast(ubyte)0u).member` of type `ubyte` to parameter `ref ubyte p`
+fail_compilation/b20011.d(36): Error: function `b20011.main.assignableByOut(out ubyte p)` is not callable using argument types `(ubyte)`
```
Comment (Basile-z) [inline test/fail_compilation/b20011.d]: The messages in my PR are better. Yours don't mention anything about constness and are rather confusing

Comment (kinke) [reply test/fail_compilation/b20011.d]: I totally disagree.

Comment (Basile-z) [reply test/fail_compilation/b20011.d]: The test cases are in now. Free to you to rework but I still think that it's sub-accurate to just output that the func cant be called. The user would have to think more about the why.

**Comment by kinke** [comment]
Reworked & ready for review. Phobos is ready too; currently still blocked by Sociomantic's ocean lib, which needs something like https://github.com/sociomantic-tsunami/ocean/pull/726.

**Comment by Geod24** [comment]
CC @matthias-wende-sociomantic

**Comment by Geod24** [comment]
This PR is blocked on a new release from your side for Ocean, Turtle, and Swarm
@ben-palmer-sociomantic @don-clugston-sociomantic

**Comment by don-clugston-sociomantic** [comment]
> This PR is blocked on a new release from your side for Ocean, Turtle, and Swarm
@ben-palmer-sociomantic @don-clugston-sociomantic

We're still blocked waiting a DMD nightly build, so that we have a compiler with both the patches we need.

**Comment by Geod24** [comment]
Because of [this](https://forum.dlang.org/post/cvhmdpmfdemfzarmzwdm@forum.dlang.org) I guess ?

**Comment by rainers** [comment]
> We're still blocked waiting a DMD nightly build, so that we have a compiler with both the patches we need.

Nightlies are available for a couple of days now.

**Comment by kinke** [comment]
With Sociomantic shut down, will someone continue to maintain Ocean and put out a new release?

**Comment by ibuclaw** [comment]
I think we should negotiate a handover to the relevant stakeholders who are wanting to continue working on the library. But getting any traction on this is like trying to boil an ocean.

As an owner of the organisation, I could simply just go ahead and do this without going through the bureaucracy process and see who complains (I expect no one).

**Comment by wilzbach** [comment]
@Geod24 I saw that you pushed to the Ocean repository recently. What's your opinion on this? Is it possible to get a new tag or do we need to drop it from Buildkite?

**Comment by Geod24** [comment]
We are working on getting Ocean back in shape so it isn't a blocker to anyone anymore.

**Comment by Geod24** [comment]
A new Ocean has been released. However, since the failure is also present in swarm and turtle, those needs a release too. I'll see if I can do the fix at the Buildkite level instead.

**Comment by kinke** [comment]
Thx Mathias, Ocean is finally green one year and a bunch of rebases later; the 2 other Sociomantic projects are still blocking this as you've mentioned. 

I have another PR with breaking changes blocked by Buildkite projects (incl. Ocean!) - #11387. I don't think the PR opener should be responsible for fixing all breakages (and wait for the according tag, and rebase to fix accumulating merge conflicts...) in order to make progress...

**Comment by Geod24** [comment]
Still working on getting the other two projects ready. Almost there.

> I don't think the PR opener should be responsible for fixing all breakages

I understand the feeling, but at the same time, I think it's good we have barriers to breaking change.

> and wait for the according tag

That was the real culprit here. Going with HEAD would have solved the issue for Ocean.

But the issue is really with the Buildkite setup: It has many (documented) flaws, but for all its flaws, it is an invaluable help. When developing patches for the compiler, I often end up opening the PR early so I can assess the impact of my changes.

Side note: I have a project that aims to fix this issue (a Github app that could automate testing at a large scale). But as always, it's not something that will happen overnight.

**Comment by MoonlightSentinel** [comment]
> A new Ocean has been released. However, since the failure is also present in swarm and turtle, those needs a release too. I'll see if I can do the fix at the Buildkite level instead.

@Geod24 How should that be implemented? Explcitly selecting the revision for the repository and it's submodules?

**Comment by Geod24** [comment]
> Explcitly selecting the revision for the repository and it's submodules?

Unless you have an easy way to say "Get the latest minor for this".

**Comment by MoonlightSentinel** [comment]
Don't think so, see dlang/ci#429

**Comment by Geod24** [comment]
Swarm released, doing Turtle soon

**Comment by Geod24** [comment]
Done, can you give this a rebase ?

**Comment by PetarKirov** [comment]
@ibuclaw can you please take a look again?

**Comment by kinke** [comment]
> can you please take a look again?

Not sure why you're asking Iain, he has neither reviewed this nor the original PR.

I've amended a testcase for 20608, a regression introduced by #10115 (reverted here).

**Comment by kinke** [comment]
Rebased once more...

**Review by ibuclaw — APPROVED** [review]
Let's go, as ICE's would also be present in gdc-master due to bad AST sent from the front-end.

**Comment by Geod24** [comment]
This broke our code. Reduced test case:
```D
public T deserializeFull (T) () @safe
{
    static if (is(T : E*, E))
    {
        return &[ deserializeFull!(typeof(T.init[0]))() ][0];
    }
    return T.init;
}

struct Block { string value; }

void main ()
{
    auto b = deserializeFull!(immutable(Block)*)();
}
```

Here, the literal is a GC allocation, so why is the compiler refusing it ? Moreover, I'm asking for an `immutable` value (originally was asking for `const`, doesn't make a difference), so the error message is at least misleading.

**Comment by ibuclaw** [comment]
> This broke our code. Reduced test case:

Thanks, can we have a regression issue?

**Comment by Geod24** [comment]
Sure. Just wanted to confirm if that was intended. Anyway, opened https://issues.dlang.org/show_bug.cgi?id=21312

**Comment by ibuclaw** [comment]
> Sure. Just wanted to confirm if that was intended. Anyway, opened https://issues.dlang.org/show_bug.cgi?id=21312

It may be intended, but I think is better to digress in a bug report.

**Comment by kinke** [comment]
This is intended. People seem to have abused this `[<rvalue>][0]` to turn an rvalue into an lvalue, surely many times without realizing that this entails creating a GC array, e.g., https://github.com/dlang/dmd/pull/10124#issuecomment-508890883.

**Comment by kinke** [comment]
It might be nice to have a little `moveToGC` helper in `core.memory` or so, for something like this without array overhead:
```D
extern (C) void* _d_newitemU(in TypeInfo _ti);

T* moveToGC(T)(auto ref T value)
{
    import core.lifetime;
    auto mem = cast(T*) _d_newitemU(typeid(T)); // allocate without initializing
    moveEmplace(value, *mem);
    return mem;
}

public T deserializeFull (T) () @safe
{
    static if (is(T : immutable E*, E))
        return () @trusted { return cast(T) deserializeFull!E.moveToGC(); }();
    else
        return T.init;
}
```

**Comment by Geod24** [comment]
Well I'd be happy to use `[ ... ].ptr`, unfortunately someone decided it wasn't `@safe`.

**Comment by Geod24** [comment]
>  but I'd still consider [<elem>].ptr obscure and an explicit moveToGC() much better, even without considering the array overhead.

"I consider this syntax obscure" is not reason enough to break the language. If it was, I would have nuked `with` statements from orbit :)
The code in question is in a serializer. Using `&[ ... ][0]` worked as intended, and now it does not, for what seems to be an arbitrary limitation. Worse it comes without a deprecation. Where is the benefit to the user ?

**Comment by kinke** [comment]
That comment was wrt. *your* code example, not in general.

The benefit IMO is consistency. Indexing or dot-var'ing a literal now consistently yields an rvalue, array literals being no exception. My thinking was that a part of literal is still a literal. E.g., this used to compile too (without `@safe`) and was the reason for the Ocean issue IIRC:
```D
struct S { int x; }
void main() { auto p = &S.init.x; *p = 123; }
```

IIRC (it's been a while), NOT doing this would lead to compile errors with master not being caught by the compiler anymore after cast-related changes/fixes in earlier commits.

**Comment by Geod24** [comment]
> The benefit IMO is consistency. Indexing or dot-var'ing a literal now consistently yields an rvalue, array literals being no exception.

I would rather have the compiler be consistent on the whole expression than just two levels of it.
The full expression includes a `&` which should *not* do this optimization unless there's also a dereference in the same expression, or the compiler has another mean to prove the address doesn't escape.

> My thinking was that a part of a literal is still a literal.

Right, but `[]` has always returned an lvalue. If we can optimize it to sometimes return rvalues it's great, but the current state seems problematic.
For example, what about:
```
const foo = [ 42, 24 ];
```
If we want to be consistent, shouldn't this give us a `const int[2]` ?

And yeah, the code you liked is obviously wrong, and I'm glad you put so much effort into fixing this. I just think we can do this fix while still retaining the old behavior when someone takes the address of an array literal.

**Comment by kinke** [comment]
I gave this some more thought yesterday. I do see that a (dynamic) array literal is a special case, and while the slice itself is an rvalue, its elements clearly are not. On the other hand, emplacing a single element in a new GC-allocation is the only use case I can think of for immediately indexing an array literal, and that comes with associated array overhead. So I think `moveToGC()` is better for this task, but comes with a limitation, as moving means the source value must be mutable.

I can open a new PR treating array literal elements as lvalues again, to check the effect on the compile errors.

**Comment by JohanEngelen** [comment]
This introduced regression https://issues.dlang.org/show_bug.cgi?id=23273

**Pull request merged**